contracts/src/CrossChainApplications/GETTING_STARTED.md

```diff
@@ -0,0 +1,214 @@
+
+```solidity
+// Receive a new message from another chain.
+function receiveTeleporterMessage(
+    bytes32 originChainID,
+    address originSenderAddress,
+    bytes calldata message
+) external {
+    // Only the Teleporter receiver can deliver a message.
+    require(msg.sender == address(teleporterMessenger), "Unauthorized.");
+
+    // do something with message.
+    return true;
+}
+```
+
+The base of sending and receiving messages cross chain is complete. `MyExampleCrossChainMessenger` can now be expanded with functionality that saves the received messages, and allows users to query for the latest message received from a specified chain.
+
+## Step 4: Storing the Message
+
+Start by defining the `struct` for how to save our messages. It saves the string message itself and the address of the sender.
+
+A map will also be added where the key is the `originChainID`, and the value is the latest `message` sent from that chain.
+
+```solidity
+// Messages sent to this contract.
+struct Message {
+    address sender;
+    string message;
+}
+
+mapping(bytes32 originChainID => Message message) private _messages;
+```
+
+Next, update `receiveTeleporterMessage` to save the message into our mapping after we receive and verify that it's sent from Teleporter. ABI decode the `message` bytes into a string.
+
+```solidity
+// Receive a new message from another chain.
+function receiveTeleporterMessage(
+    bytes32 originChainID,
+    address originSenderAddress,
+    bytes calldata message
+) external {
+    // Only the Teleporter receiver can deliver a message.
+    require(msg.sender == address(teleporterMessenger), "Unauthorized.");
+
+    // Store the message.
+    messages[originChainID] = Message(originSenderAddress, abi.decode(message, (string)));
+}
+```
+
+Next, add a function called `getCurrentMessage` that allows users or contracts to easily query our contract for the latest message sent by a specified chain.
+
+```solidity
+// Check the current message from another chain.
+function getCurrentMessage(
+    bytes32 originChainID
+) external view returns (address, string memory) {
+    Message memory messageInfo = messages[originChainID];
+    return (messageInfo.sender, messageInfo.message);
+}
+```
+
+There we have it, a simple cross chain messenger built on top of Teleporter! Full example [here](./ExampleMessenger/ExampleCrossChainMessenger.sol).
+
+## Step 5: Testing
+
+For testing, `scripts/local/test.sh` sets up a local Avalanche network with three subnets deployed with Teleporter, and a relayer to deliver Teleporter messages. To add an integration test simply add a new test script under `integration-tests`. An integration test for `ExampleCrossChainMessenger` is already included (`scripts/local/integration_tests/example_messenger.sh`), which performs the following steps:
```

Comment:
ran `scripts/local/test.sh` in teleporter base directory and it failed. 
I believe it is because while running `/code/docker/run_setup.sh` subnet-evm is not cloned to the docker container before sourcing the latest avalanchego version from `./scripts/versions.sh` in subnet-evm. 
```
ARCH set to arm64
Using published awm-relayer image
Waiting for containers to start...
WARN[0000] The "TEST_TARGET" variable is not set. Defaulting to a blank string. 
[+] Building 0.8s (22/22) FINISHED                                                                                        docker:desktop-linux
 => [test_runner internal] load build definition from Dockerfile                                                                          0.0s
 => => transferring dockerfile: 2.22kB                                                                                                    0.0s
 => [test_runner internal] load .dockerignore                                                                                             0.0s
 => => transferring context: 2B                                                                                                           0.0s
 => [test_runner internal] load metadata for docker.io/library/ubuntu:20.04                                                               0.7s
 => [local_network internal] load .dockerignore                                                                                           0.0s
 => => transferring context: 2B                                                                                                           0.0s
 => [local_network internal] load build definition from Dockerfile                                                                        0.0s
 => => transferring dockerfile: 2.22kB                                                                                                    0.0s
 => [test_runner auth] library/ubuntu:pull token for registry-1.docker.io                                                                 0.0s
 => [local_network  1/14] FROM docker.io/library/ubuntu:20.04@sha256:ed4a42283d9943135ed87d4ee34e542f7f5ad9ecf2f244870e23122f703f91c2     0.0s
 => CACHED [test_runner  2/14] RUN apt-get update &&     apt-get upgrade -y &&     apt-get clean                                          0.0s
 => CACHED [test_runner  3/14] RUN apt-get install -y wget curl git python3 make gcc build-essential                                      0.0s
 => CACHED [test_runner  4/14] RUN apt-get install -y bison                                                                               0.0s
 => CACHED [test_runner  5/14] RUN wget https://go.dev/dl/go1.20.10.linux-arm64.tar.gz                                                    0.0s
 => CACHED [test_runner  6/14] RUN rm -rf /usr/local/go && tar -C /usr/local -xzf go1.20.10.linux-arm64.tar.gz                            0.0s
 => CACHED [test_runner  7/14] RUN go version                                                                                             0.0s
 => CACHED [test_runner  8/14] RUN apt-get update && apt-get install -y locales && rm -rf /var/lib/apt/lists/*     && localedef -i en_US  0.0s
 => CACHED [test_runner  9/14] RUN curl -L https://raw.githubusercontent.com/ava-labs/foundry/v0.1.0/foundryup/install > /tmp/foundry-in  0.0s
 => CACHED [test_runner 10/14] RUN apt-get update &&     apt-get install -y python3-pip                                                   0.0s
 => CACHED [test_runner 11/14] RUN pip3 install base58                                                                                    0.0s
 => CACHED [test_runner 12/14] RUN curl -sSfL https://raw.githubusercontent.com/ava-labs/avalanche-cli/main/scripts/install.sh | sh -s -  0.0s
 => CACHED [test_runner 13/14] RUN avalanche --version                                                                                    0.0s
 => CACHED [test_runner 14/14] RUN echo '{"MetricsEnabled}":false}' > ~/.avalanche-cli/config                                             0.0s
 => [test_runner] exporting to image                                                                                                      0.0s
 => => exporting layers                                                                                                                   0.0s
 => => writing image sha256:cd2514843237abc94311537a801c1aa85072f4f2993e865baea3213adcaef35e                                              0.0s
 => => naming to docker.io/library/test-runner-image                                                                                      0.0s
 => [local_network] exporting to image                                                                                                    0.0s
 => => exporting layers                                                                                                                   0.0s
 => => writing image sha256:164396b1c78dee645d78e82b1fea8512637af7fd14e1f6fd2b3c9f2bad4d4339                                              0.0s
 => => naming to docker.io/library/local-network-image                                                                                    0.0s
[+] Running 3/0
 ✔ Container teleporter-relayer-1  Created                                                                                                0.0s 
 ✔ Container local_network_test    Created                                                                                                0.0s 
 ✔ Container test_runner           Created                                                                                                0.0s 
Attaching to local_network_test, teleporter-relayer-1, test_runner
teleporter-relayer-1  | Waiting for subnets to start up. Retry count: 
test_runner           | Waiting for subnets to start up. Retry count: 
local_network_test    | /code/docker/run_setup.sh: line 34: ./scripts/versions.sh: No such file or directory
local_network_test exited with code 1
Aborting on container exit...
[+] Stopping 3/0
 ✔ Container test_runner           Stopped                                                                                                0.0s 
 ✔ Container teleporter-relayer-1  Stopped                                                                                                0.0s 
 ✔ Container local_network_test    Stopped                                                                                                0.0s 
Waiting for containers to start...
Docker process is dead
```

Reply:
might conflict with https://github.com/ava-labs/teleporter/pull/152

Reply:
Is `teleporter/subnet-evm` populated? After cloning `teleporter`, run `git submodule update --init --recursive` to do so. 

That being said, we're planning on removing the Docker-based integration tests, as well as reworking our `subnet-evm` dependency, so this part of the README will be removed soon.

Reply:
I think this was an issue with how our foundry fork was setup. If you were running this on Mac M1, then this should be resolved now. In any case, agreed not worth debugging here.